examples/tracing/undump.py

```diff
@@ -0,0 +1,202 @@
+
+BPF_PERF_OUTPUT(unix_recv_events);
+
+//static int unix_stream_recvmsg(struct socket *sock, struct msghdr *msg,
+//			       size_t size, int flags)
+int trace_stream_entry(struct pt_regs *ctx)
+{
+    int ret = PT_REGS_RC(ctx);
+    
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    u32 pid = pid_tgid >> 32;
+    u32 tid = pid_tgid;
+
+    FILTER_PID
+
+    struct stream_data_t data4 = {.pid = pid,};
+    data4.uid = bpf_get_current_uid_gid();
+    data4.ts_us = bpf_ktime_get_ns() / 1000;
+
+    struct socket *sock = (struct socket *)PT_REGS_PARM1(ctx);   
+    struct msghdr *msg = (struct msghdr *)PT_REGS_PARM2(ctx);    
+
+    data4.sock_state = sock->state;
+    data4.sock_type = sock->type;
+    data4.sock_flags = sock->flags;
+
+    data4.msg_namelen = msg->msg_namelen;
+    
+    bpf_get_current_comm(&data4.task, sizeof(data4.task));
+    
+    struct unix_sock *unsock = (struct unix_sock *)sock->sk;
+    data4.unix_sock_path = (char *)unsock->path.dentry->d_name.name;
+    
+    stream_recvmsg_events.perf_submit(ctx, &data4, sizeof(data4));
+
+    return 0;
+};
+
+int trace_unix_stream_read_actor(struct pt_regs *ctx)
+{
+    u32 zero = 0;
+    int ret = PT_REGS_RC(ctx);
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    u32 pid = pid_tgid >> 32;
+    u32 tid = pid_tgid;
+
+    FILTER_PID
+
+	struct sk_buff *skb = (struct sk_buff *)PT_REGS_PARM1(ctx);
+
+    struct recv_data_t *data = unix_data.lookup(&zero);
+    if (!data) 
+        return 0;
+
+    unsigned int data_len = skb->len;
+    if(data_len > MAX_PKT)
+        return 0;
+        
+    void *iodata = (void *)skb->data;
+    data->recv_len = data_len;
+    
+    bpf_probe_read(data->pkt, data_len, iodata);
+    unix_recv_events.perf_submit(ctx, data, data_len+sizeof(u32));
+    
+    return 0;
+}
+"""
+
+if args.pid:
+    bpf_text = bpf_text.replace('FILTER_PID',
+        'if (pid != %s) { return 0; }' % args.pid)
+
+bpf_text = bpf_text.replace('FILTER_PID', '')
+
+# process event
+def print_stream_event(cpu, data, size):
+    event = b["stream_recvmsg_events"].event(data)
+    global start_ts
+    if args.timestamp:
+        if start_ts == 0:
+            start_ts = event.ts_us
+        printb(b"%-9.3f" % ((float(event.ts_us) - start_ts) / 1000000), nl="")
+    printb(b"%-6s %-12s" % (event.pid, event.task))
+
+# process event
+def print_recv_pkg(cpu, data, size):
+    event = b["unix_recv_events"].event(data)
+    print("----------------", end="")
+    for i in range(0, event.recv_len):
+        print("%02x " % event.pkt[i], end="")
+        sys.stdout.flush()
+        if (i+1)%16 == 0:
+            print("")
+            print("----------------", end="")
+    print("\n----------------recv %d bytes" % event.recv_len)
+    
+# initialize BPF
+b = BPF(text=bpf_text)
+b.attach_kprobe(event="unix_stream_recvmsg", fn_name="trace_stream_entry")
+b.attach_kprobe(event="unix_stream_read_actor", fn_name="trace_unix_stream_read_actor")
```

Comment:
I am curious why you need to trace two functions to capture packets. 

I think `unix_stream_recvmsg` should be enough.

Reply:
“unix_stream_recvmsg” can not get sk_buff directly, "trace_unix_stream_read_actor" is easier. but must use "unix_stream_recvmsg" get  COMM.

Reply:
Do you mean that `trace_unix_stream_read_actor` is executed under softirq context ?

Reply:
No， `unix_stream_read_actor` is easy to get sk_buff, when `recv msg` happened. 
function call stack like:
```
unix_stream_recvmsg() -----> can not get sk_buff, or hardly, but use it to get/filter COMM and  PID
   ...
      unix_stream_read_actor() -----> get sk_buff is easier than unix_stream_recvmsg.
```

Reply:
So why not just use `trace_unix_stream_read_actor` ?